Threaded motorized stage manager

kangz12345 opened the pull request:
This PR implements `StageManager` which manages several stage RPC clients and controls them in a thread (other than the main GUI thread) to prevent GUI from freezing.

I tested it working, using the following code:

```python
app = QApplication([])

thread = QThread()
manager = StageManager()
manager.moveToThread(thread)
thread.start()

manager.connectStage.emit("test", ("ip", port, "serial_number"))
proxy = StageProxy(manager, "test")

widget = StageWidget()  # will be introduced in the next PR (already implemented)
widget.moveBy.connect(proxy.moveBy)
widget.show()

app.exec_()
```

However, I forgot to save the actual test script, so there might be mistakes in the example code above.
Please don't hesitate to ask.

kangz12345 committed:
Add stage.py and `StageControllerApp`

kangz12345 committed:
Add `StageWidget`

kangz12345 committed:
Change the docstring for `moveBy` signal

kangz12345 committed:
Add signal emission of `StageWidget`

kangz12345 committed:
Add `StageControllerFrame`

kangz12345 committed:
Add `sipyco` to workflow dependency

kangz12345 committed:
Introduce `StageManager`

This object will be moved to a new thread later.

kangz12345 committed:
Remove unnecessary docstring

kangz12345 committed:
Add more description for `StageManager`

kangz12345 committed:
Implement `_connect()`

kangz12345 committed:
Setup generalized API structure of `StageManager`

kangz12345 committed:
Add `clear`

kangz12345 committed:
Add `closeTarget`

kangz12345 committed:
Change `StageManager` to `QThread`

kangz12345 committed:
Add `use_client()` decorator

kangz12345 committed:
Add description in `use_client()`

kangz12345 committed:
Handle `OSError` in `use_client()`

kangz12345 committed:
Add `moveTo` and `moveBy`

kangz12345 committed:
Fix typo

kangz12345 committed:
Fix typo

kangz12345 committed:
Fix missing `self`

kangz12345 committed:
Make `StageManager` a `QObject`

kangz12345 committed:
Add `StageProxy` for convenient signal connection

kangz12345 committed:
Add usage example docstring in `StageProxy`

kangz12345 committed:
Remove widget parts for small PR

kangz12345 committed:
Remove trailing whitespaces

kangz12345 committed:
Add module docstring

kangz12345 committed:
Move `use_client()` outside `StageManager`

kangz12345 committed:
Add type annotation

kangz12345 committed:
Add pylint disable comments

kangz12345 committed:
Add consistent return value

kangz12345 committed:
Change `cache` -> `lru_cache(maxsize=8)`

kangz12345 committed:
Add pylint disable comment

kangz12345 committed:
Remove `default` argument from `use_client()`

Wrong way to define a decorator with arguments.

kangz12345 committed:
Make 2 blank lines between global definitions

BECATRUE commented on iquip/apps/stage.py:
```diff
@@ -0,0 +1,163 @@
+        self._clients[key] = Client(*info)
+
+    @pyqtSlot(str, float)
+    @use_client
```
It looks too clear!

BECATRUE commented on iquip/apps/stage.py:
```diff
@@ -0,0 +1,163 @@
+            first argument.
+    """
+    @functools.wraps(function)
+    def wrapped(self: "StageManager", key: str, *args, **kwargs):
```
Referred to [stackoverflow](https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class), and [PEP 563](https://peps.python.org/pep-0563/#enabling-the-future-behavior-in-python-3-7), you might be able to use `__future__` module.

Please check it!

kangz12345 replied:
Good to know! Thank you, I will apply it.

BECATRUE commented on iquip/apps/stage.py:
```diff
@@ -0,0 +1,163 @@
+            return
+        try:
+            function(self, client, *args, **kwargs)
+        except OSError:
```
Why did you check only `OSError`?
According to [Client](https://github.com/m-labs/sipyco/blob/master/sipyco/pc_rpc.py#L70), there seems to be possibilities that `ValueError` and `AttributeError` are raised.

kangz12345 replied:
That's correct! I guess it would be good to catch them all. Even though `AttributeError` is the programmer's responsibility, catching it will helpful for stability.

BECATRUE commented on iquip/apps/stage.py:
```diff
@@ -0,0 +1,163 @@
+        try:
+            function(self, client, *args, **kwargs)
+        except OSError:
+            logger.exception("Error occurred while running %s with client %s.", function, key)
```
Did you intend `function`, instead of `function.__name__`?

kangz12345 replied:
Good idea, I will apply that.

BECATRUE commented on iquip/apps/stage.py:
```diff
@@ -0,0 +1,163 @@
+    def __init__(self, manager: StageManager, key: str):
+        """
+        Args:
+            manager: StageManager object where the target stage client object lives.
```
The `manager` and `key` should be moved to the attributes section.

kangz12345 committed:
Use postponed evaluation of annotations (PEP563)

Applied the review.

kangz12345 committed:
Catch `AttributeError` and `ValueError` as well

Applied the review.

kangz12345 committed:
Log function's name instead of object itself

Applied the review.

kangz12345 committed:
Fix docstring

Applied the review.

kangz12345 committed:
Move __future__ import to the top

As PEP8 says, the __future__ import line should be
right after the module docstring.

See also: https://peps.python.org/pep-0008/#module-level-dunder-names

BECATRUE reviewed (approved):
LGTM
